### DISCUSSION TITLE
Authelia two factor business logic is confusing / wrong

### DISCUSSION THREAD

**Discussion by gene1wood** [authelia/authelia#7873] [Q&A]
It seems like there are a few aspects of two factor authentication in authelia that are logically confusing or wrong. I wanted to start this as a discussion before opening a GitHub issue in case I'm missing something and this is actually correct

# To allow users to use two factor, a policy must *require* two factor

In Authelia 4.38.9, if the `configuration.yml` contains a policy that *requires* [`one_factor`](https://www.authelia.com/configuration/security/access-control/#one_factor), this indicates that the administrator doesn't want to require users to use two factor, but if users do use two factor, that's fine. The [`one_factor`](https://www.authelia.com/configuration/security/access-control/#one_factor) documentation says

> This policy requires the user at minimum complete 1FA successfully (username and password). This means if they have performed 2FA then they will be allowed to access the resource.

This makes sense.

If however the `configuration.yml` contains only policies that require `one_factor` (no policy has a requirement for `two_factor`), then authelia does not present to users the option to enroll in TOTP or WebAuthn.

This doesn't make sense to me. Just because the administrator doesn't want to *require* users to use two factor, doesn't mean that the administrator wants to *prevent* users from opting in to getting the security benefits for their account of using a second factor. It's confusing that authelia must have a policy in place *requiring* two factor in order for users to be able to opt in to using two factor on their accounts.

I think that whether or not users can opt in to using two factor should have nothing to do with whether the authelia configuration contains access control policies that *require* two factor.

# For users who have provisioned TOTP or WebAuthn, they're not prompted for it when visiting one_factor policy resources

If a user sets up two factor authentication and then visits a resource that has a policy requirement of `one_factor`, authelia doesn't prompt the user to provide their second factor.

This seems weird and different from how other websites function.

Other sites (e.g. Google, Facebook, etc) seem to follow this logic

* For all resources, if the user has only one factor configured, prompt them to enter their password
* For all resources, if the user has two factor configured, prompt them for password and their second factor

I would hope that authelia follows this type of common user flow during login and instead

* For all resources, if the authelia user has two factor configured, prompt them for password and their second factor
* If the user has only one factor (their password) configured, and they visit a resource that has a policy of `one_factor`, let the user access the resource
* If the user has only one factor (their password) configured, and they visit a resource that has a policy of `two_factor`, bar the user access and redirect them to setup two factor authentication

# When a single two factor policy exists, the main page shows a confusing message

If there are no access control policies in `configuration.yml` that *require* `two_factor`, then, when the user browses to the main page of authelia after logging in, they see a page that looks like this

![authelia_no_two_factor](https://github.com/user-attachments/assets/17a22cf4-a8b6-49c2-a03d-73d6d2a68818)

This makes sense.

If however there are any policies that *require* `two_factor`, then when the user navigates to the main page (e.g. https://authelia.example.com/ ) the user is redirected to https://authelia.example.com/2fa/one-time-password and sees a page that looks like this

![authelia_two_factor](https://github.com/user-attachments/assets/97989f01-d799-43e0-92e8-c8da8dafa4a9)

The page contains the text saying

> The resource you're attempting to access requires two-factor authentication
> Register your first device by clicking on the link below

This doesn't make sense.

The resource that the user browsed to is the main page of the authelia installation (e.g. https://authelia.example.com/ ). This page doesn't require two-factor authentication.

Why is a user who navigates to the main page of authelia redirected to https://authelia.example.com/2fa/one-time-password with a message saying that "The resource you're attempting to access requires two-factor authentication"?

It makes sense to redirect to this page when a user navigates to a resource that has an authelia access control policy that *requires* `two_factor`. It doesn't make sense to do so when the user just navigates to the main authelia page, for example to administer their user account.

Am I overlooking something or is this indeed a confusing or wrong aspect of authelia?

**Comment by james-d-elliott** [comment]
I believe this stems from a wrong understanding about the minimum. The administrator sets the minimum authentication level a user is required to have setup and used to authenticate for any given resource. The users currently registered 2FA methods does not affect this intentionally. If a user however visits a resource that only requires 1FA then they meet the minimum requirement if they are already 1FA'd or 2FA'd, thus are allowed access.

As far as requiring 2FA without context of the intended resource, this is for multiple reasons. For starters without knowledge the fact that there are resources that may require 2FA means the user may be required to provide 2FA to authenticate to their desired resource in the very near future. Secondly there are elements where it is or can be required for them to be 2FA'd before managing their own account; both currently and in the near future this will be drastically expanded.

For any setup that does NOT have a two_factor policy there is no reason to offer 2FA configuration as it would be ineffectual. It would lead to another form of confusion and has done in the past where the user/admin is unsure why they were even asked to setup 2FA if it's never used.

> For all resources, if the user has only one factor configured, prompt them to enter their password

This is a great illustration where the very relaxed security model these other solutions provide is not compatible with the ideas of Authelia. If the admin configures a resource to require 2FA, the users should always provide 2FA. Is there merit in the idea of allowing users to require more than the minimum in the future? Absolutely, this would probably be a nice to have addition in the future. 

Authelia is however a security driven project, and we already give users a more secure option than those described in this post; administrators can set granular policies where they set every resource policy to `two_factor` or `deny`, or alternatively / in addition / etc require a network element for each resource.

**Comment by gene1wood** [comment]
> For starters without knowledge the fact that there are resources that may require 2FA means the user may be required to provide 2FA to authenticate to their desired resource in the very near future.

The issue is that the current logic in authelia, does not consider whether or not a given user has access to a resource that requires 2FA. For example

If you have a resource in `configuration.yml` like this

```yaml
access_control:
  rules:
    - domain: text.example.com
      subject:
        - 'group:this-group-has-no-members'
      policy: two_factor
```

Then all users, when they navigate to the main page of authelia will be redirected to setup their two factor. This is despite the fact that there are no resources that they can access that require two factor.

> For any setup that does NOT have a two_factor policy there is no reason to offer 2FA configuration as it would be ineffectual. It would lead to another form of confusion and has done in the past where the user/admin is unsure why they were even asked to setup 2FA if it's never used.

That makes perfect sense to me. But with the current design, this exact scenario occurs if any rule is created that has a `policy: two_factor`, regardless of whether or not the user has access to that resource.

> Is there merit in the idea of allowing users to require more than the minimum in the future? Absolutely, this would probably be a nice to have addition in the future.

Agreed. Just reaching parity in security provided by Google would be great here (with Google, if a user has two factor configured, they are prompted for their second factor)

> Authelia is however a security driven project, and we already give users a more secure option than those described in this post

I'd challenge the notion that Authelia is more secure than, for example, Google's authentication system.

Thanks for sharing your thoughts on this @james-d-elliott . My read from you is that

* You'd be open to a feature request to allow administrators to turn on an option where if a user has two factor configured, they're always prompted for it, regardless of the policy for a given resource. I'll open that request.
* The behavior that I'm seeing in authelia which is, by my view, unexpected, where enabling two factor on a resource that maybe one user has access to, switches all users into a mode where they are prompted to setup two factor in the main page, is the expected behavior.

I'll open an issue suggesting an expansion of the documentation to make this behavior clear. I encountered this after reading [the two factor documentation](https://www.authelia.com/configuration/second-factor/introduction/) but finding nothing indicating how to turn two factor authentication on. Only after a lot of testing did I find that I needed to start by creating a resource, even one that had no authorized groups, that had a `policy: two_factor`.

If this sounds good, feel free to close out this discussion. Thanks again for building authelia, it's such a great product. I advocate for it's use whenever possible.

**Reply by james-d-elliott** [reply]
> That makes perfect sense to me. But with the current design, this exact scenario occurs if any rule is created that has a policy: two_factor, regardless of whether or not the user has access to that resource.

Yep it is how it works. I believe the reason we came to this decision (we need an ADR probably) is that complexity is the enemy of good design and security and the complexity required to address this with the forward auth flow is sufficiently important to address the associated risk. Unfortunately proxies like nginx are still popular as oposed to modern proxies which provide much better support for this, and we generally don't want to remove support for something as popular. It's not how it operates with OpenID Conect 1.0 however as it is effectively proxy agnostic.

> I'd challenge the notion that Authelia is more secure than, for example, Google's authentication system.

Authelia allows admins to very granularity specify the level required for a particular resource based on several attributes that identify the client and the resource, regardless of their currently enrolled 2FA methods. We offer admins the ability to customize the allowed methods for 2FA. 

Google's ability to do this is severely limited by their design and they quite literally cannot compete with us in this area; which is fine, they offer a completely different product not similar to Authelia in many of the areas that matter. But it is a legitimate reason to question their implementation being used as a figurative bible to beat us with; we don't intend to compete with any particular product offering (especially one like theirs where it's security as an after-thought not security first) and would rather judge each mechanism on the merits. 

To be fair they also have a long history of very poor security and privacy practices which most people would be aware of in some fashion. For example until earlier this year they still supported SMS 2FA which as a mechanism has had several major breaches associated directly with the poor security this practice provides.

> You'd be open to a feature request to allow administrators to turn on an option where if a user has two factor configured, they're always prompted for it, regardless of the policy for a given resource. I'll open that request.

Fairly sure there is already one but yes it's an idea that we're more than open to as it would offer an enhancement (if only marginal) to security. 

> The behavior that I'm seeing in authelia which is, by my view, unexpected, where enabling two factor on a resource that maybe one user has access to, switches all users into a mode where they are prompted to setup two factor in the main page, is the expected behavior.

Yes, we specifically do not enable 2FA if no rules that could benefit from it take effect. But as soon as any policy exists that may require it we enable it globally. Initially we had it on at all times, so if a user visited directly they'd be prompted to perform the setup. However this particular element lead to a quite a few confused people regarding why we were asking for this when it's not enabled. I can't recall anyone else who was confused by the current method.

> I'll open an issue suggesting an expansion of the documentation to make this behavior clear. I encountered this after reading [the two factor documentation](https://www.authelia.com/configuration/second-factor/introduction/) but finding nothing indicating how to turn two factor authentication on. Only after a lot of testing did I find that I needed to start by creating a resource, even one that had no authorized groups, that had a policy: two_factor.

Issues are for bugs or feature requests specifically when they are software related; documentation changes can either just be PR'd (it's all markdown and very easy to do) or made a discussion.
